### PR TITLE
Feature/render titles with custom id markdown

### PR DESCRIPTION
**Pull request by m-murasovs**

### CODE REVIEW

File: test/markdown_renderers.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                    name="welcome-to-apify"
+                    href="#welcome-title-id"
+                    id="#welcome-title-id">
+                    <span class="header-link"></span>
+                </a>
+                Welcome to Apify
+            </h1>`);
+    });
+
+    it('returns a header without ID param when no ID is passed', () => {
```
Comment (mtrunkat) [inline test/markdown_renderers.js:44]: When there is no ID specified it should generate it from the heading text to be backward compatible

File: test/markdown_renderers.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    const renderer = new marked.Renderer();
+    renderer.heading = customHeadingRenderer;
+    marked.use({ renderer });
+
+    it('returns a header with ID param when ID is passed', () => {
+        const renderedTitle = function () {
+            return marked('# Welcome to Apify {#welcome-title-id}');
+        };
+        expect(renderedTitle()).to.equal(`
+            <h1>
```
Comment (mtrunkat) [inline test/markdown_renderers.js:17]: The format should be slightly different:

```
<h2 id="container-web-server">
    <a href="#container-web-server"></a>
    Container web server
</h2>
```

File: test/markdown_renderers.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                <a 
+                    name="welcome-to-apify"
+                    href="#welcome-title-id"
+                    id="#welcome-title-id">
```
Comment (mtrunkat) [inline test/markdown_renderers.js:21]: There is no hash in ID

File: test/markdown_renderers.js (hunk 2 above)
Comment (mtrunkat) [reply test/markdown_renderers.js:17]: You don't need a name

File: src/markdown_renderers.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+export const customHeadingRenderer = (text, level) => {
+    let headingToReturn;
+
+    const idRegEx = /[^{}]+(?=})/g;
+    const idTags = text.match(idRegEx);
+
+    if (idTags) {
+        const idTag = formatIdTag(idTags[0]);
```
Comment (mtrunkat) [inline src/markdown_renderers.js:26]:
```
if (idTag !== idTags[0]) log.error('Badly formatted heading ID', { id: idTags[0] });
```

File: src/markdown_renderers.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        idTag = idTag.substring(0, idTag.length - 1);
+    }
+    return idTag;
+}
+
+export const customHeadingRenderer = (text, level) => {
+    let headingToReturn;
+
+    const idRegEx = /[^{}]+(?=})/g;
+    const idTags = text.match(idRegEx);
```
Comment (mtrunkat) [inline src/markdown_renderers.js:23]:
```
const idTags = text.match(idRegEx);
const idTag = idTags && idTags.length ? formatIdTag(idTags[0]) : text;
```

Comment (mtrunkat) [reply src/markdown_renderers.js:23]:
```
const idTags = text.match(idRegEx);
let idTag = idTags && idTags.length ? formatIdTag(idTags[0]) : null;
if (!idTag) idTag = formatIdTag(text);


``

File: src/markdown_renderers.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import log from './log';
+
+function formatIdTag(idTag) {
+    // Get rid of whitespace and random characters
+    idTag = idTag.toLowerCase().trim().replace(/[^\w]+/g, '-');
+    // Remove dashes at the start
+    while (idTag[0] === '-') {
```
Comment (jancurn) [inline src/markdown_renderers.js:7]: This looks bit dirty, you can replace this with a simple regex like:

```
idTag.replace(/^[\-]+|[\-]+$/g, '');
```

File: src/markdown_renderers.js [hunk 7]
```diff
@@ -4,9 +4,7 @@ function formatIdTag(idTag) {
     // Get rid of whitespace and random characters
     idTag = idTag.toLowerCase().trim().replace(/[^\w]+/g, '-');
     // Remove dashes at the start
-    while (idTag[0] === '-') {
-        idTag = idTag.replace(idTag[0], '');
-    }
+    idTag = idTag.replace(/^[-]+|[-]+$/g, '');
```
Comment (jancurn) [inline src/markdown_renderers.js:7]: This covers also the "Remove trailing dashes" case :)

Comment (m-murasovs) [reply src/markdown_renderers.js:7]: :)) thank you!

**Pull request merged**